internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+            s"Standard API name change in modified class.",
+            usedNames = usedNames
+          )
+      }
+    }
+
+    private final var initial: Option[zprof.InitialChanges] = None
+    def registerInitial(changes: InitialChanges): Unit = {
+      import scala.collection.JavaConverters._
+      val fileChanges = changes.internalSrc
+      val profChanges = zprof.Changes(
+        added = toStringTableIndices(toPathStrings(fileChanges.getAdded.asScala)).toList,
+        removed = toStringTableIndices(toPathStrings(fileChanges.getRemoved.asScala)).toList,
+        modified = toStringTableIndices(toPathStrings(fileChanges.getChanged.asScala)).toList
+      )
+      initial = Some(
+        zprof.InitialChanges(
+          changes = Some(profChanges),
+          removedProducts = toStringTableIndices(toPathStrings(changes.removedProducts)).toList,
+          binaryDependencies = toStringTableIndices(toPathStrings(changes.binaryDeps)).toList,
+          externalChanges = toApiChanges(changes.external).toList
+        )
+      )
+    }
+
+    private final var currentEvents: List[zprof.InvalidationEvent] = Nil
+    def registerEvent(
+        kind: String,
+        inputs: Iterable[String],
+        outputs: Iterable[String],
+        reason: String
+    ): Unit = {
+      val event = zprof.InvalidationEvent(
+        kind = kind,
+        inputs = toStringTableIndices(inputs).toList,
+        outputs = toStringTableIndices(outputs).toList,
+        reason = reason
+      )
+
+      currentEvents = event :: currentEvents
+    }
+
+    private final var cycles: List[zprof.CycleInvalidation] = Nil
+    def registerCycle(
+        invalidatedClasses: Iterable[String],
+        invalidatedPackageObjects: Iterable[String],
+        initialSources: Iterable[File],
+        invalidatedSources: Iterable[File],
+        recompiledClasses: Iterable[String],
+        changesAfterRecompilation: APIChanges,
+        nextInvalidations: Iterable[String],
+        shouldCompileIncrementally: Boolean
+    ): Unit = {
+      val newCycle = zprof.CycleInvalidation(
+        invalidated = toStringTableIndices(invalidatedClasses).toList,
+        invalidatedByPackageObjects = toStringTableIndices(invalidatedPackageObjects).toList,
+        initialSources = toStringTableIndices(toPathStrings(initialSources)).toList,
+        invalidatedSources = toStringTableIndices(toPathStrings(invalidatedSources)).toList,
+        recompiledClasses = toStringTableIndices(recompiledClasses).toList,
+        changesAfterRecompilation = toApiChanges(changesAfterRecompilation).toList,
+        nextInvalidations = toStringTableIndices(nextInvalidations).toList,
+        startTimeNanos = compilationStartNanos,
+        compilationDurationNanos = compilationDurationNanos,
+        events = currentEvents,
+        shouldCompileIncrementally = shouldCompileIncrementally
+      )
+
+      cycles = newCycle :: cycles
+      ()
+    }
+  }
+}
+
+abstract class RunProfiler {
```

Comment:
Same here :wink: I find some scaladocs on a fresh API usually very helpful to get
the intention and purpose of the API. The question _why is this interface necessary_
should be answered in the scaladocs :smiley: